docs/instrumentation.md

```diff
@@ -0,0 +1,283 @@
+instrumentation of a specific module, allowing for tracing the execution
+of module functions.
+
+This could be especially useful in the process of development of different
+protocol backends in our `lpntpn` like LPC or I2C.
+
+For now, scripts are located inside our fork of
+[ms-tpm-20-ref](https://github.com/lpn-plant/ms-tpm-20-ref) repository.
+Look for `Samples/Nucleo-TPM/scripts/kfunc_tracer` directory on the dev branch.
+
+We use the integrated functionality that gcc provides '-finstrument-function'
+for function instrumentation.
+This step allows us to register exact function call addresses and log them to
+dmesg.
+
+Next, we match the function addresses against kernel processes address space
+that's located in `/proc/modules` to identify loaded module, that fn call
+belongs to. By subtracting function address with start memory location of loaded
+module, we obtain the real offset of a function call, that could be easily
+matched with an exact function name.
+
+The last step is to extract functions memory offsets located in non-striped
+kernel modules. For this purpose we are using a simple readelf call:
+`readelf -s MODULE_NAME.ko | grep FUNC`.
+
+This process requires kernel recompilation, execution of module startup and
+parsing of resulting trace files.
+
+### Prepare Raspberry Pi
+
+We are using Raspberry Pi 3, but with minor modifications, this process could
+be used on different targets.
+
+**Flash SD card with raspbian lite os.**
+
+At the moment of writing image `2021-05-07-raspios-buster-armhf-lite.img` was
+used.
+
+**Enable ssh.**
+
+Execute on raspberry pi with connected keyboard and display
+
+```shell
+# upgrade
+sudo apt update
+sudo apt upgrade
+
+# change default password
+passwd
+
+# enable ssh
+sudo systemctl enable ssh
+sudo systemctl start ssh
+```
+
+**Check ssh connection.**
+
+Environmental variable `RASPI_IP` will be used several more times during this
+guide.
+
+Run on your host machine to test ssh connection.
+```shell
+export RASPI_IP=pi@192.168.8.170
+ssh-copy-id -i ~/.ssh/id_rsa.pub $RASPI_IP
+ssh $RASPI_IP echo "success"
+# should echo success without prompting user to enter password
+```
+
+### Kernel compinaltion
+
+**Prepare build environment.**
+
+Ubuntu 20.04 LTS was used in this example.
+
+Install dependencies and get source code.
+```shell
+sudo apt install crossbuild-essential-armhf     # default toolchain
+sudo apt install python-is-python3              # no python2 in ubuntu 20 fix
+sudo apt install bison flex
+sudo apt install qtbase5-dev pkg-config         # xconfig
+sudo apt install libncurses5-dev                # menuconfig
+
+cd git
+git clone –-depth=1 https://github.com/raspberrypi/linux
+cd linux
+```
+
+**Patch kernel**
+
+For future debug output we add `CONFIG_DYNAMIC_DEBUG=y` to bcm2709_defconfig.
+The rest of this patch is the code instrumentation functions themself.
+
+[instrument.patch](data/instrument.patch)
+
+**Build kernel**
+
+```shell
+cd git
```

Comment:
@acwiek Sometimes (in other documents) you use `$` as a prompt before command. Please be consistent and either use it always or never. I'm in favor of the first one. What is more, you can use different prompts (for RPi and Host) to make it clear on which host a given command should be executed.

Reply:
That's true, I'm using `$` when there is one or more lines of output from the command.
When there is no output I'm not using this prefix for ease of copy-pasting.
According to your suggestions, I will add an extra `$` prefix for console/shell code blocks

Reply:
Added to my to-do list, I will add this change after the current PR's gets merged.